### PR TITLE
feat: native early stopping with validation, CLI exposure, and in-memory best weights

### LINKED ISSUE #39: [Python] Early Stopping with Best Model Checkpoint
## Description

Training currently runs for a fixed number of epochs regardless of model performance. Implementing early stopping will automatically halt training when the model stops improving, preventing overfitting and saving computation time.

## Current Behavior

```python
model.train(epochs=100, lr=0.01)
# Runs all 100 epochs even if loss plateaus at epoch 20
```

## Expected Behavior

```python
model.train(
    epochs=100, 
    lr=0.01,
    early_stopping=True,
    patience=10,           # Stop if no improvement for 10 epochs
    restore_best=True      # Restore weights from best epoch
)
# Stops at epoch 30 if loss stopped improving at epoch 20
# Model weights are from epoch 20 (best), not epoch 30 (last)
```

## Implementation Details

1. Add `early_stopping`, `patience`, and `restore_best` parameters to `Model.train()`
2. Track validation/training loss history during training
3. Implement patience counter - stop training if loss doesn't improve for `patience` consecutive epochs
4. Cache model weights at best epoch and restore them when training stops
5. Print informative message when early stopping triggers

## Acceptance Criteria

- [ ] Training stops automatically when loss stops improving
- [ ] Best model weights are restored (not the last epoch weights)
- [ ] Works for both classification and regression tasks
- [ ] Backward compatible - default behavior unchanged if `early_stopping=False`

## Related Issues

- #22 (Validation Split) - Early stopping works best with validation loss tracking

### PR DESCRIPTION
**Pull request by reddyeswaranush**
## Fixes
Closes #39

## Type of Change
- [ ] Bug fix
- [x] New feature
- [ ] Documentation / Refactor
- [ ] Math / Logic correction

## Description
This PR implements **native early stopping inside the Rust core** to eliminate Python–Rust per-epoch overhead.  
The best model weights are now stored and restored **in memory (RAM)** instead of using disk I/O.  
Validation set support has been added so early stopping monitors validation loss instead of training loss.  
The Python API and CLI have been updated to expose `early_stopping` and `patience` to end users.  
Additionally, a bias accumulation bug in the Linear layer forward pass has been fixed.

## How Has This Been Tested?
- [x] **Manual Testing**: Trained on the Adult dataset with a validation split and verified early stopping behavior.
- [x] **Integration**: Built Rust core with `maturin develop` and verified Python API + Rust engine integration.
- [x] **Logic Verification**: Confirmed best weights are restored when patience is exceeded.

## Contribution Context
- [x] I am contributing through the SWOC program.

### CODE REVIEW

**Comment by debug-soham** [comment]
@reddyeswaranush, please resolve the merge conflicts.

**Comment by debug-soham** [comment]
@reddyeswaranush, the build tests for this PR are failing. Please inform us of the status within 24 hours. If the issues aren't addressed by then, we will proceed to close this PR and reassign it to another contributor.

**Comment by reddyeswaranush** [comment]
Hi @debug-soham ,

I’ve spent some time working on this issue, but I’m currently unable to find a correct and clean solution.

To avoid blocking progress, I think it’s best to unassign me so someone else can take it forward.  
Thanks for the opportunity, and sorry for the inconvenience.
